plugins/single_plugins/fast-switcher.cpp

```diff
@@ -8,23 +8,56 @@
 #include <wayfire/view.hpp>
 #include <wayfire/workspace-manager.hpp>
 #include <wayfire/util/log.hpp>
+#include <wayfire/render-manager.hpp>
+#include <wayfire/util/duration.hpp>
 
 /*
  * This plugin provides abilities to switch between views.
  * It works similarly to the alt-esc binding in Windows or GNOME
  */
 
+class animated_alpha_transformer : public wf::scene::view_2d_transformer_t
+{
+  public:
+    animated_alpha_transformer(wayfire_view view,
+        std::shared_ptr<wf::config::option_t<int>> length) : wf::scene::view_2d_transformer_t(view), anim(
+            length, wf::animation::smoothing::sigmoid)
+    {
+        alpha = 1.0;
+    }
+
+    void start(double end_alpha)
+    {
+        anim.animate(alpha, end_alpha);
+    }
+
+    void step()
+    {
+        if (!anim.running())
+        {
+            return;
+        }
+
+        alpha = static_cast<double>(anim);
+        view->damage();
+    }
+
+    wf::animation::simple_animation_t anim;
+};
+
 class wayfire_fast_switcher : public wf::per_output_plugin_instance_t, public wf::keyboard_interaction_t
 {
     wf::option_wrapper_t<wf::keybinding_t> activate_key{"fast-switcher/activate"};
     wf::option_wrapper_t<wf::keybinding_t> activate_key_backward{
         "fast-switcher/activate_backward"};
     wf::option_wrapper_t<double> inactive_alpha{"fast-switcher/inactive_alpha"};
+    wf::option_wrapper_t<int> animation_duration{"fast-switcher/animation_duration"};
     std::vector<wayfire_view> views; // all views on current viewport
     size_t current_view_index = 0;
     // the modifiers which were used to activate switcher
     uint32_t activating_modifiers = 0;
     bool active = false;
+    bool animations_ongoing = false;
```

Comment:
This variable seems to serve no real purpose.

Reply:
That's not quite correct: I imagine you mean that it's redundant with `active` , but it's actually not. That's because when exiting the switcher (for example, after pressing `esc` or activating a view), `active` is set to false, but the animations are still being performed for X milliseconds, so we need to keep the hooks on. This variable indicates whether the blur effect is on or not. 

In addition, if the user re-activates the switcher during that duration, we should not re-add the hooks a second time, since they're already active.

Reply:
> That's not quite correct: I imagine you mean that it's redundant with `active` , but it's actually not. 

I do not mean that it is redundant with active. I mean that you could remove this variable in every place that it is used and the code would behave the same. Let me explain. At first, when the keybinding is activated, you check if `!animations_ongoing` and set it to true, when hooking up the pre and post callbacks. Then in post callback, you assert `animations_ongoing` and then unset it when unhooking pre and post. The act of adding and removing the pre/post effects is enough and the variable can be omitted.

Reply:
>I think what you are trying to avoid is adding multiple effects of the same type while animation is in flight? If so, perhaps this makes sense.

Yes exactly, that's what I mentioned in the second part of my comment.